pkg/controllers/cnset/resource.go

```diff
@@ -0,0 +1,200 @@
+		},
+	}
+	return cnCloneSet
+}
+
+func syncPersistentVolumeClaim(cn *v1alpha1.CNSet, cloneSet *kruise.CloneSet) {
+	dataPVC := corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: dataVolume,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+			Resources: corev1.ResourceRequirements{
+				Requests: map[corev1.ResourceName]resource.Quantity{
+					corev1.ResourceStorage: cn.Spec.CacheVolume.Size,
+				},
+			},
+			StorageClassName: cn.Spec.CacheVolume.StorageClassName,
+		},
+	}
+	tpls := []corev1.PersistentVolumeClaim{dataPVC}
+	cn.Spec.Overlay.AppendVolumeClaims(&tpls)
+	cloneSet.Spec.VolumeClaimTemplates = tpls
+}
+
+func syncReplicas(cn *v1alpha1.CNSet, cs *kruise.CloneSet) {
+	cs.Spec.Replicas = &cn.Spec.Replicas
+
+}
+
+func syncPodMeta(cn *v1alpha1.CNSet, cs *kruise.CloneSet) {
+	cn.Spec.Overlay.OverlayPodMeta(&cs.Spec.Template.ObjectMeta)
+}
+
+func syncPodSpec(cn *v1alpha1.CNSet, cs *kruise.CloneSet) {
+	main := corev1.Container{
+		Name:      v1alpha1.ContainerMain,
+		Image:     cn.Spec.Image,
+		Resources: cn.Spec.Resources,
+		Command: []string{
+			"/bin/sh", fmt.Sprintf("%s/%s", configPath, Entrypoint),
+		},
+		VolumeMounts: []corev1.VolumeMount{
+			{Name: dataVolume, ReadOnly: true, MountPath: dataPath},
+			{Name: configVolume, ReadOnly: true, MountPath: configPath},
+		},
+		Env: []corev1.EnvVar{
+			util.FieldRefEnv(common.PodNameEnvKey, "metadata.name"),
+			util.FieldRefEnv(common.NamespaceEnvKey, "metadata.namespace"),
+			util.FieldRefEnv(common.PodIPEnvKey, "status.podIP"),
+			{Name: common.HeadlessSvcEnvKey, Value: utils.GetHeadlessSvcName(cn)},
+		},
+	}
+	cn.Spec.Overlay.OverlayMainContainer(&main)
+	podSpec := corev1.PodSpec{
+		Containers: []corev1.Container{main},
+		ReadinessGates: []corev1.PodReadinessGate{{
+			ConditionType: pub.InPlaceUpdateReady,
+		}},
+	}
+	common.SyncTopology(cn.Spec.TopologyEvenSpread, &podSpec)
+
+	cn.Spec.Overlay.OverlayPodSpec(&podSpec)
+	cs.Spec.Template.Spec = podSpec
+}
+
+func buildCNSetConfigMap(cn *v1alpha1.CNSet) (*corev1.ConfigMap, error) {
+	configMapName := utils.GetConfigName(cn)
+	dsCfg := cn.Spec.Config
+	// detail: https://github.com/matrixorigin/matrixone/blob/main/pkg/dnservice/cfg.go
+	if dsCfg == nil {
+		dsCfg = v1alpha1.NewTomlConfig(map[string]interface{}{
+			"service-type":   common.CNService,
+			"listen-address": listenAddress,
+			"log": map[string]interface{}{
+				"level":    logLevel,
+				"format":   logFormatType,
+				"max-size": logMaxSize,
+			},
+			"file-service": map[string]interface{}{
+				"backend": backendType,
+				"s3": map[string]interface{}{
+					"endpoint":   "",
+					"bucket":     "",
+					"key-prefix": "",
+				},
+			},
+			"pipeline": map[string]interface{}{
+				"host-size":     hostSize,
+				"guest-size":    guestSize,
+				"operator-size": operatorSize,
+				"batch-row":     batchRow,
+				"batch-size":    batchSize,
+			},
```

Comment:
Configuration items should be divided clearly, a configuration item is either configured by user via CR spec or controlled by operator via automation logic.